src/compiler/scala/tools/nsc/classpath/DirectoryClassPath.scala

```diff
@@ -334,7 +334,7 @@ case class DirectoryClassPath(dir: File) extends JFileDirectoryLookup[ClassFileE
 
   protected def createFileEntry(file: AbstractFile): ClassFileEntryImpl = ClassFileEntryImpl(file)
   protected def isMatchingFile(f: File, siblingExists: String => Boolean): Boolean =
-    f.isClass && !(f.getName.endsWith(".class") && siblingExists(f.getName.dropRight(6) + ".tasty"))
+    f.isClass && !(f.getName.endsWith(".class") && siblingExists(classNameToTasty(f.getName)))
```

Comment:
How about testing both cases? I.e., if the classfile is `X$.class`, drop it if there is either `X.class` or `X$.class`.

Reply:
Right, the extra lookups are bad.

IIUC, the compiler will anyway find the `A$.tasty` file and load it. Doing so should re-use an already existing symbol for `A$` that could be there from loading the `.class` file.